src/signals-unix.c

```diff
@@ -225,6 +225,18 @@ static void segv_handler(int sig, siginfo_t *info, void *context)
     jl_ptls_t ptls = jl_get_ptls_states();
     assert(sig == SIGSEGV || sig == SIGBUS);
 
+    // if we're profiling, this segfault is likely caused by the unwinder.
+    // ignore the signal and jump back to where we came from.
+    if (running && ptls->safe_restore) {
```

Comment:
Why do you need this? Does it not work with the condition below?

Reply:
Just being conservative, only affecting the case where the profiler is running. Do it unconditionally then?

Reply:
I mean you shouldn't need any code here in the segfault handler. Have you tested that it doesn't work without this but with the condition a few lines below?

Reply:
(Or in another word, it is meant to be doing this unconditionally)

Reply:
Ah ok, no it doesn't work, looking closer it triggers a segfault in `jl_call_in_ctx` (via `jl_throw_in_ctx..., jl_stackovf_exception, ...)`).
Inferring from the function names, doesn't that behave differently from the plain `longjmp` I do here, would I need to catch an exception then somehow?

Reply:
Hmm, segfault **in** `jl_call_in_ctx`? Did you get a `NULL` `context`?

Reply:
Ah, I guess this thread doesn't have `signal_stack` allocated. I believe this should fix it (you can merge this with the falllback `ifdef` below if you want).

```patch
diff --git a/src/signals-unix.c b/src/signals-unix.c
index 0fafe121cd..8da89b5fc4 100644
--- a/src/signals-unix.c
+++ b/src/signals-unix.c
@@ -89,6 +89,14 @@ static void jl_call_in_ctx(jl_ptls_t ptls, void (*fptr)(void), int sig, void *_c
     // checks that the syscall is made in the signal handler and that
     // the ucontext address is valid. Hopefully the value of the ucontext
     // will not be part of the validation...
+    if (!ptls->signal_stack) {
+        sigset_t sset;
+        sigemptyset(&sset);
+        sigaddset(&sset, sig);
+        sigprocmask(SIG_UNBLOCK, &sset, NULL);
+        fptr();
+        return;
+    }
     uintptr_t rsp = (uintptr_t)ptls->signal_stack + sig_stack_size;
     assert(rsp % 16 == 0);
 #if defined(_OS_LINUX_) && defined(_CPU_X86_64_)
```

Reply:
We don't handle segfault the same way on OSX. I don't really know if the two ways could be used together.